src/main/java/io/vertx/core/eventbus/impl/clustered/ClusteredEventBus.java

```diff
@@ -206,31 +193,18 @@ public MessageImpl createMessage(boolean send, String address, MultiMap headers,
   protected <T> void sendOrPub(OutboundDeliveryContext<T> sendContext) {
     if (((ClusteredMessage) sendContext.message).getRepliedTo() != null) {
       clusteredSendReply(((ClusteredMessage) sendContext.message).getRepliedTo(), sendContext);
+    } else if (sendContext.options.isLocalOnly()) {
+      super.sendOrPub(sendContext);
     } else {
-      if (sendContext.options.isLocalOnly()) {
-        super.sendOrPub(sendContext);
-      } else if (Vertx.currentContext() != sendContext.ctx) {
-        // Current event-loop might be null when sending from non vertx thread
-        sendContext.ctx.runOnContext(v -> {
-          subs.get(sendContext.message.address(), ar -> onSubsReceived(ar, sendContext));
-        });
-      } else {
-        subs.get(sendContext.message.address(), ar -> onSubsReceived(ar, sendContext));
-      }
-    }
-  }
-
-  private <T> void onSubsReceived(AsyncResult<ChoosableIterable<ClusterNodeInfo>> asyncResult, OutboundDeliveryContext<T> sendContext) {
-    if (asyncResult.succeeded()) {
-      ChoosableIterable<ClusterNodeInfo> serverIDs = asyncResult.result();
-      if (serverIDs != null && !serverIDs.isEmpty()) {
-        sendToSubs(serverIDs, sendContext);
-      } else {
-        super.sendOrPub(sendContext);
-      }
-    } else {
-      log.error("Failed to send message", asyncResult.cause());
-      sendContext.written(asyncResult.cause());
+      sendContext.ctx.succeededFuture(sendContext.message)
+        .flatMap(deliveryStrategy::chooseNodes)
+        .onFailure(t -> {
```

Comment:
avoid usage of `onFailure`/`onSuccess` for efficiency purpose

Reply:
Done